### PR TITLE
TLS for Raft Join

### PR DESCRIPTION
**Pull request by vishalnayak**

### CODE REVIEW

File: helper/testhelpers/testhelpers.go [hunk 1]
```diff
@@ -767,7 +766,7 @@ func RaftClusterJoinNodes(t testing.T, cluster *vault.TestCluster) {
 	{
 		core := cluster.Cores[1]
 		core.UnderlyingRawStorage.(*raft.RaftBackend).SetServerAddressProvider(addressProvider)
-		_, err := core.JoinRaftCluster(namespace.RootContext(context.Background()), leaderAPI, leaderCore.TLSConfig, false)
+		_, err := core.JoinRaftCluster(namespace.RootContext(context.Background()), leaderCore.Client, false)
```
Comment (briankassouf) [inline helper/testhelpers/testhelpers.go]: I don't think we want to use this client here as it does retries, which is not ideal for Joining

File: http/sys_raft.go [hunk 2]
```diff
@@ -27,7 +29,31 @@ func handleSysRaftJoinPost(core *vault.Core, w http.ResponseWriter, r *http.Requ
 		return
 	}
 
-	joined, err := core.JoinRaftCluster(context.TODO(), req.LeaderAddr, nil, req.Retry)
+	if req.LeaderAddr == "" {
+		respondError(w, http.StatusBadRequest, errors.New("leader api address is required"))
+		return
+	}
+
+	config := api.DefaultConfig()
```
Comment (briankassouf) [inline http/sys_raft.go]: Could we keep the custom client code from the Join function, it is making it so there are no retries

File: http/sys_raft.go [hunk 3]
```diff
@@ -27,7 +29,31 @@ func handleSysRaftJoinPost(core *vault.Core, w http.ResponseWriter, r *http.Requ
...
+	config.Address = req.LeaderAddr
+	if req.CACert != "" && req.ClientCert != "" && req.ClientKey != "" {
+		tlsConfig := &api.TLSConfig{
+			CACert:     req.CACert,
+			ClientCert: req.ClientCert,
+			ClientKey:  req.ClientKey,
+		}
+		if err := config.ConfigureTLS(tlsConfig); err != nil {
```
Comment (briankassouf) [inline http/sys_raft.go]: I believe this only works with paths to Cert/Key files. instead we want to provide the full config (path wont exist on the server)

File: helper/testhelpers/testhelpers.go (hunk 1 above)
Comment (vishalnayak) [reply helper/testhelpers/testhelpers.go]: Fixed.

File: http/sys_raft.go (hunk 2 above)
Comment (vishalnayak) [reply http/sys_raft.go]: Fixed.

File: http/sys_raft.go (hunk 3 above)
Comment (vishalnayak) [reply http/sys_raft.go]: Fixed.

File: http/sys_raft.go [hunk 4]
```diff
@@ -27,7 +31,40 @@ func handleSysRaftJoinPost(core *vault.Core, w http.ResponseWriter, r *http.Requ
 		return
 	}
 
-	joined, err := core.JoinRaftCluster(context.Background(), req.LeaderAddr, nil, req.Retry)
+	var tlsConfig *tls.Config
+	switch {
+	case req.LeaderCACert != "" && req.LeaderClientCert != "" && req.LeaderClientKey != "":
+		// Create TLS Config
+		pool := x509.NewCertPool()
```
Comment (jefferai) [inline http/sys_raft.go]: I approved this but if you can move this section to sdk/helper/tlsutil we can reuse it elsewhere pretty quickly...

Comment (vishalnayak) [reply http/sys_raft.go]: I'll do this in a separate PR.

**Pull request merged**